// angular-tinymce/bundles/angular-tinymce.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('rxjs/Observable'), require('rxjs/add/operator/debounceTime'), require('rxjs/add/operator/skipWhile'), require('rxjs/add/operator/take'), require('rxjs/add/observable/interval'), require('rxjs/add/observable/fromEvent'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/forms', 'rxjs/Observable', 'rxjs/add/operator/debounceTime', 'rxjs/add/operator/skipWhile', 'rxjs/add/operator/take', 'rxjs/add/observable/interval', 'rxjs/add/observable/fromEvent', '@angular/common'], factory) :
	(factory((global['angular-tinymce'] = {}),global.ng.core,global.ng.forms,global.Rx,global.Rx.Observable.prototype,global.Rx.Observable.prototype,global.Rx.Observable.prototype,global.Rx.Observable,global.Rx.Observable,global.ng.common));
}(this, (function (exports,core,forms,Observable,debounceTime,skipWhile,take,interval,fromEvent,common) { 'use strict';

var TinyMceEvents = (function () {
    function TinyMceEvents() {
    }
    return TinyMceEvents;
}());
TinyMceEvents.Click = 'click';
TinyMceEvents.DblClick = 'dblClick';
TinyMceEvents.MouseDown = 'MouseDown';
TinyMceEvents.MouseUp = 'MouseUp';
TinyMceEvents.MouseMove = 'MouseMove';
TinyMceEvents.MouseOver = 'MouseOver';
TinyMceEvents.MouseOut = 'MouseOut';
TinyMceEvents.MouseEnter = 'MouseEnter';
TinyMceEvents.MouseLeave = 'MouseLeave';
TinyMceEvents.KeyDown = 'KeyDown';
TinyMceEvents.KeyPress = 'KeyPress';
TinyMceEvents.KeyUp = 'KeyUp';
TinyMceEvents.ContextMenu = 'ContextMenu';
TinyMceEvents.Paste = 'Paste';
TinyMceEvents.Init = 'Init';
TinyMceEvents.Focus = 'Focus';
TinyMceEvents.Blur = 'Blur';
TinyMceEvents.BeforeSetContent = 'BeforeSetContent';
TinyMceEvents.SetContent = 'SetContent';
TinyMceEvents.GetContent = 'GetContent';
TinyMceEvents.PreProcess = 'PreProcess';
TinyMceEvents.PostProcess = 'PostProcess';
TinyMceEvents.NodeChange = 'NodeChange';
TinyMceEvents.Undo = 'Undo';
TinyMceEvents.Redo = 'Redo';
TinyMceEvents.Change = 'Change';
TinyMceEvents.Dirty = 'Dirty';
TinyMceEvents.Remove = 'Remove';
TinyMceEvents.ExecCommand = 'ExecCommand';
TinyMceEvents.PastePreProcess = 'PastePreProcess';
TinyMceEvents.PastePostProcess = 'PastePostProcess';
TinyMceEvents.AddEditor = 'AddEditor';
TinyMceEvents.RemoveEditor = 'RemoveEditor';
var TINYMCE_SETTINGS_TOKEN = new core.InjectionToken('angular-tinymce-settings');
var TinyMceComponent = (function () {
    /**
     * @param {?} _input_settings
     * @param {?} ngZone
     */
    function TinyMceComponent(_input_settings, ngZone) {
        var _this = this;
        this._input_settings = _input_settings;
        this.ngZone = ngZone;
        this.click = new core.EventEmitter();
        this.dblclick = new core.EventEmitter();
        this.mousedown = new core.EventEmitter();
        this.mouseup = new core.EventEmitter();
        this.mousemove = new core.EventEmitter();
        this.mouseover = new core.EventEmitter();
        this.mouseout = new core.EventEmitter();
        this.mouseenter = new core.EventEmitter();
        this.mouseleave = new core.EventEmitter();
        this.keydown = new core.EventEmitter();
        this.keypress = new core.EventEmitter();
        this.keyup = new core.EventEmitter();
        this.contextmenu = new core.EventEmitter();
        this.paste = new core.EventEmitter();
        this.init = new core.EventEmitter();
        this.focus = new core.EventEmitter();
        this.blur = new core.EventEmitter();
        this.beforesetcontent = new core.EventEmitter();
        this.setcontent = new core.EventEmitter();
        this.getcontent = new core.EventEmitter();
        this.preprocess = new core.EventEmitter();
        this.postprocess = new core.EventEmitter();
        this.nodechange = new core.EventEmitter();
        this.undo = new core.EventEmitter();
        this.redo = new core.EventEmitter();
        this.change = new core.EventEmitter();
        this.dirty = new core.EventEmitter();
        this.remove = new core.EventEmitter();
        this.execcommand = new core.EventEmitter();
        this.pastepreprocess = new core.EventEmitter();
        this.pastepostprocess = new core.EventEmitter();
        this.onModelChange = function () { };
        this.onModelTouched = function () { };
        this._setSettings(this.settings);
        if (!window.tinymce && !document.getElementById('tinymceScript')) {
            var tag = document.createElement('script');
            tag.id = 'tinymceScript';
            tag.setAttribute('src', this.settings.tinymceScriptURL || 'assets/tinymce/tinymce.min.js');
            tag.onload = function () {
                tinymce.baseURL = _this.settings.baseURL;
            };
            document.body.appendChild(tag);
        }
    }
    Object.defineProperty(TinyMceComponent.prototype, "settings", {
        /**
         * @return {?}
         */
        get: function () {
            return this._settings;
        },
        /**
         * @param {?} value
         * @return {?}
         */
        set: function (value) {
            if (value) {
                this._settings = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} obj
     * @return {?}
     */
    TinyMceComponent.prototype.writeValue = function (obj) {
        var /** @type {?} */ val = obj != null ? obj.toString() : '';
        if (this.editor) {
            this.fromWriteValue = true;
            this.editor.setContent(val);
        }
        else {
            this.beforeInitValue = val;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TinyMceComponent.prototype.registerOnChange = function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TinyMceComponent.prototype.registerOnTouched = function (fn) {
        this.onModelTouched = fn;
    };
    /**
     * @param {?} disabled
     * @return {?}
     */
    TinyMceComponent.prototype.setDisabledState = function (disabled) {
        this.disabled = disabled;
        this.setEditorMode(disabled);
    };
    /**
     * @param {?} disabled
     * @return {?}
     */
    TinyMceComponent.prototype.setEditorMode = function (disabled) {
        if (this.editor) {
            if (disabled) {
                this.editor.setMode('readonly');
            }
            else {
                this.editor.setMode('design');
            }
        }
    };
    /**
     * @param {?} settings
     * @return {?}
     */
    TinyMceComponent.prototype._setSettings = function (settings) {
        var /** @type {?} */ localSettings = settings || this._input_settings || {};
        this.settings = Object.assign({}, localSettings);
        if (((window)).tinymce) {
            tinymce.baseURL = ((this.settings)).baseURL;
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    TinyMceComponent.prototype.ngOnChanges = function (changes) {
        var /** @type {?} */ needReinit = false;
        if (changes === null) {
            return;
        }
        if (changes['settings']) {
            if (changes['settings'].currentValue) {
                this._setSettings(changes['settings'].currentValue);
                needReinit = true;
            }
        }
        if (changes['isDisabled']) {
            this.setDisabledState(changes['isDisabled'].currentValue);
        }
        if (needReinit) {
            this.reInitEditor();
        }
    };
    /**
     * @return {?}
     */
    TinyMceComponent.prototype.ngAfterViewInit = function () {
        this.initEditor();
    };
    /**
     * @return {?}
     */
    TinyMceComponent.prototype.initEditor = function () {
        var _this = this;
        this.settings.target = this.elem.nativeElement;
        this.initCallbacks(this.settings);
        Observable.Observable.interval(300)
            .skipWhile(function () { return !((window)).tinymce; })
            .take(1)
            .subscribe(function () {
            tinymce.init(_this.settings);
        });
    };
    /**
     * @return {?}
     */
    TinyMceComponent.prototype.reInitEditor = function () {
        var _this = this;
        if (this.editor) {
            this.ngZone.run(function () {
                _this.triggerChange(true);
            });
        }
        this.removeEditor();
        this.initEditor();
    };
    /**
     * @param {?} settings
     * @return {?}
     */
    TinyMceComponent.prototype.initCallbacks = function (settings) {
        var _this = this;
        var /** @type {?} */ origSetup = settings.setup;
        settings.setup = function (editor) {
            editor.on(TinyMceEvents.Init, function (e) { return _this.init.emit(e); });
            if (origSetup) {
                origSetup(editor);
            }
        };
        var /** @type {?} */ origInstanceCallback = settings.init_instance_callback;
        settings.init_instance_callback = function (editor) {
            _this.editor = editor;
            _this.setEditorMode(_this.disabled);
            if (_this.beforeInitValue != null) {
                _this.editor.setContent(_this.beforeInitValue);
            }
            if (origInstanceCallback) {
                origInstanceCallback(editor);
            }
            editor.on(TinyMceEvents.Click, function (e) { return _this.click.emit(e); });
            editor.on(TinyMceEvents.DblClick, function (e) { return _this.dblclick.emit(e); });
            editor.on(TinyMceEvents.MouseDown, function (e) { return _this.mousedown.emit(e); });
            editor.on(TinyMceEvents.MouseUp, function (e) { return _this.mouseup.emit(e); });
            editor.on(TinyMceEvents.MouseMove, function (e) { return _this.mousemove.emit(e); });
            editor.on(TinyMceEvents.MouseOver, function (e) { return _this.mouseover.emit(e); });
            editor.on(TinyMceEvents.MouseOut, function (e) { return _this.mouseout.emit(e); });
            editor.on(TinyMceEvents.MouseEnter, function (e) { return _this.mouseenter.emit(e); });
            editor.on(TinyMceEvents.MouseLeave, function (e) { return _this.mouseleave.emit(e); });
            editor.on(TinyMceEvents.KeyDown, function (e) { return _this.keydown.emit(e); });
            editor.on(TinyMceEvents.KeyPress, function (e) { return _this.keypress.emit(e); });
            editor.on(TinyMceEvents.KeyUp, function (e) {
                _this.ngZone.run(function () {
                    _this.triggerChange();
                });
                _this.keyup.emit(e);
            });
            editor.on(TinyMceEvents.ContextMenu, function (e) { return _this.contextmenu.emit(e); });
            editor.on(TinyMceEvents.Paste, function (e) { return _this.paste.emit(e); });
            editor.on(TinyMceEvents.Focus, function (e) { return _this.focus.emit(e); });
            editor.on(TinyMceEvents.Blur, function (e) { return _this.blur.emit(e); });
            editor.on(TinyMceEvents.BeforeSetContent, function (e) { return _this.beforesetcontent.emit(e); });
            editor.on(TinyMceEvents.SetContent, function (e) {
                _this.ngZone.run(function () {
                    _this.triggerChange();
                });
                _this.setcontent.emit(e);
            });
            editor.on(TinyMceEvents.GetContent, function (e) { return _this.getcontent.emit(e); });
            editor.on(TinyMceEvents.PreProcess, function (e) { return _this.preprocess.emit(e); });
            editor.on(TinyMceEvents.PostProcess, function (e) { return _this.postprocess.emit(e); });
            editor.on(TinyMceEvents.NodeChange, function (e) { return _this.nodechange.emit(e); });
            editor.on(TinyMceEvents.Undo, function (e) { return _this.undo.emit(e); });
            editor.on(TinyMceEvents.Redo, function (e) { return _this.redo.emit(e); });
            editor.on(TinyMceEvents.Change, function (e) {
                _this.ngZone.run(function () {
                    _this.triggerChange();
                });
                _this.change.emit(e);
            });
            editor.on(TinyMceEvents.Dirty, function (e) { return _this.dirty.emit(e); });
            editor.on(TinyMceEvents.Remove, function (e) { return _this.remove.emit(e); });
            editor.on(TinyMceEvents.ExecCommand, function (e) {
                _this.ngZone.run(function () {
                    _this.triggerChange();
                });
                _this.execcommand.emit(e);
            });
            editor.on(TinyMceEvents.PastePreProcess, function (e) { return _this.pastepreprocess.emit(e); });
            editor.on(TinyMceEvents.PastePostProcess, function (e) { return _this.pastepostprocess.emit(e); });
        };
    };
    /**
     * @param {?=} forReInit
     * @return {?}
     */
    TinyMceComponent.prototype.triggerChange = function (forReInit) {
        if (forReInit === void 0) { forReInit = false; }
        if (this.fromWriteValue) {
            this.fromWriteValue = false;
        }
        else {
            var /** @type {?} */ content = this.editor.getContent();
            if (!content) {
                content = '';
            }
            if (forReInit) {
                this.beforeInitValue = content;
            }
            this.onModelChange(content);
            this.onModelTouched();
        }
    };
    /**
     * @return {?}
     */
    TinyMceComponent.prototype.removeEditor = function () {
        if (this.editor) {
            ((tinymce)).remove(this.editor);
        }
    };
    /**
     * @return {?}
     */
    TinyMceComponent.prototype.ngOnDestroy = function () {
        this.removeEditor();
    };
    return TinyMceComponent;
}());
TinyMceComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'angular-tinymce',
                template: "<textarea #tinymce ></textarea>",
                providers: [{
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TinyMceComponent; }),
                        multi: true
                    }]
            },] },
];
/**
 * @nocollapse
 */
TinyMceComponent.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: [TINYMCE_SETTINGS_TOKEN,] },] },
    { type: core.NgZone, },
]; };
TinyMceComponent.propDecorators = {
    'isDisabled': [{ type: core.Input },],
    'settings': [{ type: core.Input },],
    'selector': [{ type: core.Input },],
    'click': [{ type: core.Output },],
    'dblclick': [{ type: core.Output },],
    'mousedown': [{ type: core.Output },],
    'mouseup': [{ type: core.Output },],
    'mousemove': [{ type: core.Output },],
    'mouseover': [{ type: core.Output },],
    'mouseout': [{ type: core.Output },],
    'mouseenter': [{ type: core.Output },],
    'mouseleave': [{ type: core.Output },],
    'keydown': [{ type: core.Output },],
    'keypress': [{ type: core.Output },],
    'keyup': [{ type: core.Output },],
    'contextmenu': [{ type: core.Output },],
    'paste': [{ type: core.Output },],
    'init': [{ type: core.Output },],
    'focus': [{ type: core.Output },],
    'blur': [{ type: core.Output },],
    'beforesetcontent': [{ type: core.Output },],
    'setcontent': [{ type: core.Output },],
    'getcontent': [{ type: core.Output },],
    'preprocess': [{ type: core.Output },],
    'postprocess': [{ type: core.Output },],
    'nodechange': [{ type: core.Output },],
    'undo': [{ type: core.Output },],
    'redo': [{ type: core.Output },],
    'change': [{ type: core.Output },],
    'dirty': [{ type: core.Output },],
    'remove': [{ type: core.Output },],
    'execcommand': [{ type: core.Output },],
    'pastepreprocess': [{ type: core.Output },],
    'pastepostprocess': [{ type: core.Output },],
    'elem': [{ type: core.ViewChild, args: ['tinymce',] },],
};
/**
 * @return {?}
 */
function tinymceDefaultSettings() {
    return {
        baseURL: '/assets/tinymce',
        tinymceScriptURL: '/assets/tinymce/tinymce.min.js',
        theme_url: '/assets/tinymce/themes/modern/theme.min.js',
        skin_url: '/assets/tinymce/skins/lightgray'
    };
}
/**
 * Modul für ein TinyMce Input
 */
var TinyMceModule = (function () {
    function TinyMceModule() {
    }
    /**
     * @param {?} settings
     * @return {?}
     */
    TinyMceModule.forRoot = function (settings) {
        return {
            ngModule: TinyMceModule,
            providers: [
                { provide: TINYMCE_SETTINGS_TOKEN, useValue: settings }
            ]
        };
    };
    return TinyMceModule;
}());
TinyMceModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [common.CommonModule],
                declarations: [TinyMceComponent],
                exports: [TinyMceComponent]
            },] },
];
/**
 * @nocollapse
 */
TinyMceModule.ctorParameters = function () { return []; };

exports.TinyMceComponent = TinyMceComponent;
exports.TINYMCE_SETTINGS_TOKEN = TINYMCE_SETTINGS_TOKEN;
exports.tinymceDefaultSettings = tinymceDefaultSettings;
exports.TinyMceEvents = TinyMceEvents;
exports.TinyMceModule = TinyMceModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
